addons/mail/models/mail_activity.py

```diff
@@ -572,53 +585,161 @@ def action_open_document(self):
+            aggregated data or not.
+        :return dict: {'activity_types': list of activity type info as list
+                            (id, name, mail_template, keep_done),
+                       'activity_res_ids': list<int> of record id ordered by closest date
+                            (deadline for ongoing activities, and done date for done activities),
+                       'grouped_activities': dict<dict<dict>>
+                            res_id -> activity_type_id -> dict of aggregated info
+                            where the aggregate information structured as:
+                                count_by_state: dict mapping state to count (ex.: 'planned': 2)
+                                ids: list of activity ids for the res_id and activity_type_id
+                                o_closest_date: closest date (deadline if ongoing, done date otherwise)
+                                state: aggregated state of the activities
+                                user_ids_ordered_by_deadline: activity responsible ordered by closest
+                                    deadline of the related activity
+                       }
+        """
+        user_tz = self.user_id.sudo().tz
+        Model = self.env[res_model]
+
+        # 1. Retrieve all ongoing and completed activities according to the parameters
+        activity_types = self.env['mail.activity.type'].search([('res_model', 'in', (res_model, False))],
+                                                               order='sequence, id')
+        fetch_done = fetch_done and activity_types.filtered('keep_done')
         activity_domain = [('res_model', '=', res_model)]
-        if domain or limit or offset:
-            res = self.env[res_model].search(domain or [], limit=limit, offset=offset)
-            activity_domain.append(('res_id', 'in', res.ids))
-        grouped_activities = self.env['mail.activity']._read_group(
+        is_filtered = domain or limit or offset
+        domain_res_ids = Model._search(domain or [], offset, limit) if is_filtered else []
+        if is_filtered:
+            activity_domain.append(('res_id', 'in', domain_res_ids))
+        all_activities = self.env['mail.activity'].with_context(active_test=not fetch_done).search_read(
             activity_domain,
-            ['res_id', 'activity_type_id'],
-            ['id:array_agg', 'date_deadline:min', '__count'])
-        # filter out unreadable records
-        if not domain:
-            res_ids = tuple(res_id for res_id, *_ in grouped_activities)
-            res_ids_set = set(self.env[res_model].search([('id', 'in', res_ids)])._ids)
-            grouped_activities = [a for a in grouped_activities if a[0] in res_ids_set]
+            ['activity_type_id', 'res_id', 'date_deadline', 'user_id', 'active'] +
+            (['attachment_ids'] if fetch_done else []),
+        )
+        all_ongoing_activities, all_completed_activities = partition(lambda a: a['active'], all_activities)
```

Comment:
You tend to use long variable names (but the naming is correct so that's cool :) ), here the "all" prefix does not seem necessary ? It eases reading as there are less "generic" word to analyze... like
```suggestion
        ongoing_activities, completed_activities = partition(lambda a: a['active'], all_activities)
```